### ISSUE TITLE
hmac does not allow for use of sha512

### ISSUE THREAD

**Issue by dignifiedquire** [digitalbazaar/forge#401]
Is there a reason `sha512` is not added to the supported list here:  https://github.com/digitalbazaar/forge/blob/master/js/md.js#L21?

I'm trying to support it using `hmac.start('sha512', secret)`, but because it's not added in there I can't use it.

Thanks for the help!

**Comment by dlongley** [comment]
It adds itself when you include the file (if you're using the minified version of the library it should already be included):

https://github.com/digitalbazaar/forge/blob/master/js/sha512.js#L17-L22

**Comment by dignifiedquire** [comment]
I'm afraid with the current release on npm this is not working:


```js
'use strict'

const forge = require('node-forge')

const types = ['md5', 'sha256', 'sha512']

types.forEach((type) => {
  const hmac = forge.hmac.create()
  hmac.start(type, 'hello')
  hmac.update('cool')

  console.log('digest', hmac.digest().toHex())
})

// Prints

// digest 172250d98c0076eae79364c56f999afc
// digest 74017b794a5cb906d06519dceee703905a377771ef2315c6631cb9821ec37f9f
// /Users/dignifiedquire/opensource/ipfs/js-libp2p-crypto/node_modules/node-forge/js/hmac.js:54
//           throw new Error('Unknown hash algorithm "' + md + '"');
//           ^

// Error: Unknown hash algorithm "sha512"
//     at Object.ctx.start (/Users/dignifiedquire/opensource/ipfs/js-libp2p-crypto/node_modules/node-forge/js/hmac.js:54:17)
//     at types.forEach (/Users/dignifiedquire/opensource/ipfs/js-libp2p-crypto/fail.js:9:8)
//     at Array.forEach (native)
//     at Object.<anonymous> (/Users/dignifiedquire/opensource/ipfs/js-libp2p-crypto/fail.js:7:7)
//     at Module._compile (module.js:541:32)
//     at Object.Module._extensions..js (module.js:550:10)
//     at Module.load (module.js:456:32)
//     at tryModuleLoad (module.js:415:12)
//     at Function.Module._load (module.js:407:3)
//     at Function.Module.runMain (module.js:575:10)
```

with `node-forge@0.6.39` and `node@6.0.0`.

**Comment by dlongley** [comment]
@dignifiedquire,

That does seem like a bug to me. Here's the workaround until that gets fixed:

```js
const forge = require('node-forge')

const types = ['md5', 'sha256', forge.md.sha512.create()]

types.forEach((type) => {
  const hmac = forge.hmac.create()
  hmac.start(type, 'hello')
  hmac.update('cool')

  console.log('digest', hmac.digest().toHex())
})
```

**Comment by dignifiedquire** [comment]
Thanks, this works great!

**Comment by dlongley** [comment]
Addressed by #456.

**Comment by davidlehn** [comment]
Further fixed by: https://github.com/digitalbazaar/forge/commit/dc980d8465fca06ec1ef97bf92f6024bbcb51452
